semgrep-core/src/engine/Specialize_formula.ml

```diff
@@ -25,97 +25,56 @@ type sformula =
 
-(* this will be adjusted later in range_to_pattern_match_adjusted *)
-let fake_rule_id (id, str) =
-  { PM.id = string_of_int id; pattern_string = str; message = "" }
-
-let match_selector ?err:(msg = "no match") (sel_opt : selector option) :
-    RM.ranges =
-  let get_match (x : RP.times RP.match_result) = x.matches in
-  match sel_opt with
-  | None -> failwith msg
-  | Some selector ->
-      List.map RM.match_result_to_range
-        (get_match (Lazy.force selector.lazy_matches))
-
-let select_from_ranges file (sel_opt : selector option) (ranges : RM.ranges) :
-    RM.ranges =
-  let pattern_match_from_binding selector (mvar, mval) =
-    match Visitor_AST.range_of_any_opt (M.mvalue_to_any mval) with
-    | None -> None
-    | Some range_loc ->
-        Some
-          {
-            PM.rule_id = fake_rule_id (selector.pid, fst selector.pstr);
-            PM.file;
-            PM.range_loc;
-            PM.tokens = lazy (M.ii_of_mval mval);
-            PM.env = [ (mvar, mval) ];
-          }
-  in
-  let select_from_range range =
-    match sel_opt with
-    | None -> Some range
-    | Some selector -> (
-        match
-          List.find_opt
-            (fun (mvar, _mval) -> M.equal_mvar selector.mvar mvar)
-            range.RM.mvars
-        with
-        | None -> None
-        | Some binding ->
-            pattern_match_from_binding selector binding
-            |> Common.map_opt RM.match_result_to_range)
-  in
-  Common.map_filter select_from_range ranges
-
 let selector_equal s1 s2 = s1.mvar = s2.mvar
 
 (*****************************************************************************)
 (* Converter *)
 (*****************************************************************************)
 
-let selector_from_formula match_func f =
+let selector_from_formula f =
   match f with
   | R.Leaf (R.P ({ pat = Sem (pattern, _); pid; pstr }, None)) -> (
       match pattern with
       | G.E { e = G.N (G.Id ((mvar, _), _)); _ } ->
-          Some
-            {
-              mvar;
-              pid;
-              pstr;
-              lazy_matches = lazy (match_func [ (pattern, pid, fst pstr) ]);
-            }
+          Some { mvar; pattern; pid; pstr }
       | _ -> None)
   | _ -> None
 
-let formula_to_sformula match_func formula =
+let formula_to_sformula formula =
+  let rec remove_selectors (selector, acc) formulas =
+    match formulas with
+    | [] -> (selector, acc)
+    | x :: xs ->
+        let selector, acc =
+          match (selector, selector_from_formula x) with
+          | None, None -> (None, x :: acc)
+          | Some s, None -> (Some s, x :: acc)
+          | None, Some s -> (Some s, acc)
+          | Some s1, Some s2 when selector_equal s1 s2 -> (Some s1, acc)
+          | Some s1, Some _s2 ->
+              (* patterns:
+               * ...
+               * - pattern: $X
+               * - pattern: $Y
+               *)
+              (* TODO: Should we fail here or just reported as a warning? This
```

Comment:
We should definitely catch this with the metachecker. Ideally I think this should be a warning, because while it's probably a mistake, it's a well-formed pattern. There is a hypothetical world where we do more translations of the user's rule so that we enter this state without them writing it. However, this is unlikely and hypothetical, so for now this pattern is always probably a mistake, so if failing is easier to implement I would also be fine with it.

Reply:
I tried to report the warning/error here and got into some difficulties to do it in a quick and clean way. I'll postpone it but I won't forget about it (c.f., #3547).